kernel/names.ml

```diff
@@ -551,24 +570,28 @@ module KerPair = struct
   module Self_Hashcons =
     struct
       type t = kernel_pair
-      type u = KerName.t -> KerName.t
+      type u = KerName.t Hashcons.hfun
+      (** Hash-consing (despite having the same user part implies having
+          the same canonical part is a logical invariant of the system, it
+          is not necessarily an invariant in memory, so we treat kernel
+          names as they are syntactically for hash-consing) *)
       let hashcons hkn = function
-        | Same kn -> Same (hkn kn)
-        | Dual (knu,knc) -> make (hkn knu) (hkn knc)
+        | Same kn ->
+          let kn, h = hkn kn in
+          Same kn, h
+        | Dual (knu, knc) ->
+          let knu, hu = hkn knu in
+          let knc, hc = hkn knc in
+          if KerName.equal knu knc then
```

Comment:
Since this is immediately after hashconsing could we use physical equality?

Reply:
... maybe. For now I just want to bench this to see if this rough POC has any observable effect (which I hope).